framework/src/variables/MooseVariableFV.C

```diff
@@ -0,0 +1,354 @@
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "MooseVariableFV.h"
+#include <typeinfo>
+#include "TimeIntegrator.h"
+#include "NonlinearSystemBase.h"
+#include "DisplacedSystem.h"
+#include "Assembly.h"
+
+registerMooseObject("MooseApp", MooseVariableFVReal);
+
+template <typename OutputType>
+InputParameters
+MooseVariableFV<OutputType>::validParams()
+{
+  return MooseVariableField<OutputType>::validParams();
+}
+
+template <typename OutputType>
+MooseVariableFV<OutputType>::MooseVariableFV(const InputParameters & parameters)
+  : MooseVariableField<OutputType>(parameters)
+{
+  _element_data = libmesh_make_unique<MooseVariableDataFV<OutputType>>(
+      *this, _sys, _tid, Moose::ElementType::Element, this->_assembly.elem());
+  _neighbor_data = libmesh_make_unique<MooseVariableDataFV<OutputType>>(
+      *this, _sys, _tid, Moose::ElementType::Neighbor, this->_assembly.neighbor());
+}
+
+template <typename OutputType>
+const std::set<SubdomainID> &
+MooseVariableFV<OutputType>::activeSubdomains() const
+{
+  return this->_sys.system().variable(_var_num).active_subdomains();
+}
+
+template <typename OutputType>
+Moose::VarFieldType
+MooseVariableFV<OutputType>::fieldType() const
+{
+  if (std::is_same<OutputType, Real>::value)
+    return Moose::VarFieldType::VAR_FIELD_STANDARD;
+  else if (std::is_same<OutputType, RealVectorValue>::value)
+    return Moose::VarFieldType::VAR_FIELD_VECTOR;
+  else if (std::is_same<OutputType, RealEigenVector>::value)
+    return Moose::VarFieldType::VAR_FIELD_ARRAY;
+  else
+    mooseError("Unknown variable field type");
+}
+
+template <typename OutputType>
+bool
+MooseVariableFV<OutputType>::activeOnSubdomain(SubdomainID subdomain) const
+{
+  return this->_sys.system().variable(_var_num).active_on_subdomain(subdomain);
+}
+
+template <typename OutputType>
+void
+MooseVariableFV<OutputType>::clearDofIndices()
+{
+  _element_data->clearDofIndices();
+}
+
+template <typename OutputType>
+typename MooseVariableFV<OutputType>::OutputData
+MooseVariableFV<OutputType>::getElementalValue(const Elem * elem, unsigned int idx) const
+{
+  return _element_data->getElementalValue(elem, Moose::Current, idx);
+}
+
+template <typename OutputType>
+typename MooseVariableFV<OutputType>::OutputData
+MooseVariableFV<OutputType>::getElementalValueOld(const Elem * elem, unsigned int idx) const
+{
+  return _element_data->getElementalValue(elem, Moose::Old, idx);
+}
+
+template <typename OutputType>
+typename MooseVariableFV<OutputType>::OutputData
+MooseVariableFV<OutputType>::getElementalValueOlder(const Elem * elem, unsigned int idx) const
+{
+  return _element_data->getElementalValue(elem, Moose::Older, idx);
+}
+
+template <typename OutputType>
+void
+MooseVariableFV<OutputType>::insert(NumericVector<Number> & residual)
+{
+  _element_data->insert(residual);
```

Comment:
I will really appreicate if we have some comments here.

`residual vector ` is the same `vector` we send it back to PETSc?

If yes, we might like to reconsider this design since `residual` is a moving target. `MooseVariable` is often updated during a `loop`, and in the same `loop`, `residual` will be updated likely.

Reply:
This is an identical API to MooseVariableFE. I doubt @rwcarlsen even uses this method for finite volume. If there's a problem it needs to be addressed in MooseVariableFE

Reply:
We should probably just remove it from the FV side.

Reply:
This is used for initial conditions FWIW.  And we can't delete this function even for FV :-(